libsqlite3-sys/build.rs

```diff
@@ -369,4 +519,128 @@ mod bindings {
         file.write_all(output.as_bytes())
             .unwrap_or_else(|_| panic!("Could not write to {:?}", out_path));
     }
+
+    #[cfg(feature = "loadable_extension")]
+    fn get_struct_by_name(bindgen_sources: &str, name: &str) -> Option<syn::ItemStruct> {
+        let file = syn::parse_file(&bindgen_sources).expect("unable to parse early bindgen output");
+
+        for item in &file.items {
+            if let syn::Item::Struct(s) = item {
+                if s.ident == name {
+                    return Some(s.to_owned());
+                }
+            }
+        }
+        None
+    }
+
+    #[cfg(feature = "loadable_extension")]
+    fn bare_fn_from_type_path(t: &syn::Type) -> syn::TypeBareFn {
+        let path = match t {
+            syn::Type::Path(tp) => &tp.path,
+            _ => {
+                panic!("type was not a type path");
+            }
+        };
+
+        let mut path_args: Option<syn::PathArguments> = None;
+        for segment in &path.segments {
+            if segment.arguments.is_empty() {
+                continue;
+            }
+            path_args = Some(segment.arguments.to_owned());
+            break;
+        }
+        match path_args {
+            Some(syn::PathArguments::AngleBracketed(p)) => {
+                for gen_arg in p.args {
+                    match gen_arg {
+                        syn::GenericArgument::Type(syn::Type::BareFn(bf)) => {
+                            return bf;
+                        }
+                        _ => {
+                            panic!("parsed type was not a bare function as expected");
+                        }
+                    };
+                }
+            }
+            _ => {
+                panic!("parsed path args were not angle bracketed as expected");
+            }
+        };
+        panic!("unexpected failure to parse bare function");
+    }
+
+    #[cfg(feature = "loadable_extension")]
+    fn generate_wrapper(
+        field_ident: &syn::Ident,
+        syn_type: &syn::Type,
+        api_fn_name: &str,
+    ) -> String {
+        use quote::quote;
+        use syn::Token;
+
+        let field_name = field_ident.to_string();
+        let api_fn_ident = syn::Ident::new(&api_fn_name, field_ident.span());
+
+        // add wrapper macro invocation to be appended to the generated bindings
+        let bare_fn = bare_fn_from_type_path(syn_type);
+        let api_fn_output = &bare_fn.output;
+
+        // prepare inputs
+        let mut api_fn_inputs = bare_fn.inputs.clone();
+
+        // handle variadic api functions
+        if bare_fn.variadic.is_some() {
+            // until rust c_variadic support exists, we can't
```

Comment:
This isn't really safe. Vararg calling convention isn't guaranteed to be the same as the fixed arg calling convention on all platforms / compilers (https://www.gnu.org/software/libc/manual/html_node/Calling-Variadics.html mentions this).

That said I guess it should be fine in practice, but still is worrisome...

Reply:
Discussion moved to https://github.com/rusqlite/rusqlite/pull/910/files#r765322613